sdk/include/sdk/middleware/Middleware.h

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2023 Robert Bosch GmbH
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef VEHICLE_APP_SDK_MIDDLEWARE_MIDDLEWARE_H
+#define VEHICLE_APP_SDK_MIDDLEWARE_MIDDLEWARE_H
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+namespace velocitas {
+
+class IPubSubClient;
+
+/**
+ * @brief Abstract base class for concrete middleware implementations
+ */
+class Middleware {
+public:
+    /**
+     * @brief Returns a reference to a singelton instance of a concrete middleware class
+     *
+     * @return Middleware&
+     */
+    static Middleware& getInstance() {
+        static std::unique_ptr<Middleware> singleton = instantiate();
+        return *singleton;
+    }
+
+    static std::string getTypeDefiningEnvVarName();
```

Comment:
Any reason why this is a static method rathern than a constant on header level?

Reply:
No clue, why I did it as a function. The definition I prefer to have in the cpp-file.